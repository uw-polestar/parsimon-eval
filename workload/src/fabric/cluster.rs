use std::iter;

use parsimon::core::network::{
    types::{Link, Node},
    NodeId,
};
use rustc_hash::FxHashMap;

pub const NR_RACKS_PER_POD: usize = 48;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Cluster {
    pub planes: Vec<Plane>,
    pub pods: Vec<Pod>,
    pub fab2spine: Vec<Link>,
}

impl Cluster {
    pub fn nr_pods(&self) -> usize {
        self.pods.len()
    }

    pub fn nr_tors_per_pod(&self) -> usize {
        self.pods.get(0).map(|p| p.racks.len()).unwrap_or(0)
    }

    pub fn nodes(&self) -> impl Iterator<Item = &Node> {
        self.planes
            .iter()
            .flat_map(|pl| pl.iter())
            .chain(self.pods.iter().flat_map(|p| p.nodes()))
    }

    pub fn links(&self) -> impl Iterator<Item = &Link> {
        self.fab2spine
            .iter()
            .chain(self.pods.iter().flat_map(|p| p.links()))
    }

    pub fn contiguousify(&mut self) {
        // Collect nodes in a very specific way: hosts first, then ToRs, the fabric switches, then
        // spine switches, preserving all ordering.
        let spines = self.planes.iter().flat_map(|plane| plane.iter());
        let mut fabs = Vec::new();
        let mut tors = Vec::new();
        let mut hosts = Vec::new();
        for pod in &self.pods {
            for fab in &pod.fabs {
                fabs.push(fab);
            }
            for rack in &pod.racks {
                tors.push(&rack.tor);
                for host in &rack.hosts {
                    hosts.push(host);
                }
            }
        }
        let nodes = hosts
            .into_iter()
            .chain(tors.into_iter().chain(fabs.into_iter().chain(spines)));

        // Now rename node IDs.
        let old2new = nodes
            .into_iter()
            .enumerate()
            .map(|(i, n)| (n.id, NodeId::new(i)))
            .collect::<FxHashMap<_, _>>();
        for plane in &mut self.planes {
            for spine in plane {
                rename_node(spine, &old2new);
            }
        }
        for pod in &mut self.pods {
            pod.rename(&old2new);
        }
        for link in &mut self.fab2spine {
            rename_link(link, &old2new);
        }
    }
}

pub type Plane = Vec<Node>;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Pod {
    pub fabs: Vec<Node>,
    pub racks: Vec<Rack>,
    pub tor2fab: Vec<Link>,
}

impl Pod {
    pub fn nodes(&self) -> impl Iterator<Item = &Node> {
        self.fabs
            .iter()
            .chain(self.racks.iter().flat_map(|r| r.nodes()))
    }

    pub fn links(&self) -> impl Iterator<Item = &Link> {
        self.tor2fab
            .iter()
            .chain(self.racks.iter().flat_map(|r| r.links()))
    }

    fn rename(&mut self, old2new: &FxHashMap<NodeId, NodeId>) {
        for fab in &mut self.fabs {
            rename_node(fab, old2new);
        }
        for rack in &mut self.racks {
            rack.rename(old2new);
        }
        for link in &mut self.tor2fab {
            rename_link(link, old2new);
        }
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Rack {
    pub tor: Node,
    pub hosts: Vec<Node>,
    pub host2tor: Vec<Link>,
}

impl Rack {
    pub fn nodes(&self) -> impl Iterator<Item = &Node> {
        iter::once(&self.tor).chain(self.hosts.iter())
    }

    pub fn links(&self) -> impl Iterator<Item = &Link> {
        self.host2tor.iter()
    }

    fn rename(&mut self, old2new: &FxHashMap<NodeId, NodeId>) {
        rename_node(&mut self.tor, old2new);
        for host in &mut self.hosts {
            rename_node(host, old2new);
        }
        for link in &mut self.host2tor {
            rename_link(link, old2new);
        }
    }
}

fn rename_node(node: &mut Node, old2new: &FxHashMap<NodeId, NodeId>) {
    node.id = *old2new.get(&node.id).unwrap();
}

fn rename_link(link: &mut Link, old2new: &FxHashMap<NodeId, NodeId>) {
    link.a = *old2new.get(&link.a).unwrap();
    link.b = *old2new.get(&link.b).unwrap();
}

#[cfg(test)]
mod tests {
    use crate::testing::TINY_CLUSTER;

    use super::*;

    #[test]
    fn contiguousify_correct() -> anyhow::Result<()> {
        let mut cluster: Cluster = serde_json::from_str(TINY_CLUSTER)?;
        cluster.contiguousify();
        insta::assert_yaml_snapshot!(&cluster);
        Ok(())
    }
}
