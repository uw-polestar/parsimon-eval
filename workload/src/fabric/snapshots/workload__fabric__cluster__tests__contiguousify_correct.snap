---
source: workload/src/fabric/cluster.rs
expression: "&cluster"
---
planes:
  - - id: 16
      kind: Switch
    - id: 17
      kind: Switch
  - - id: 18
      kind: Switch
    - id: 19
      kind: Switch
pods:
  - fabs:
      - id: 12
        kind: Switch
      - id: 13
        kind: Switch
    racks:
      - tor:
          id: 8
          kind: Switch
        hosts:
          - id: 0
            kind: Host
          - id: 1
            kind: Host
        host2tor:
          - a: 0
            b: 8
            bandwidth: 10000000000
            delay: 1000
          - a: 1
            b: 8
            bandwidth: 10000000000
            delay: 1000
      - tor:
          id: 9
          kind: Switch
        hosts:
          - id: 2
            kind: Host
          - id: 3
            kind: Host
        host2tor:
          - a: 2
            b: 9
            bandwidth: 10000000000
            delay: 1000
          - a: 3
            b: 9
            bandwidth: 10000000000
            delay: 1000
    tor2fab:
      - a: 8
        b: 12
        bandwidth: 40000000000
        delay: 1000
      - a: 8
        b: 13
        bandwidth: 40000000000
        delay: 1000
      - a: 9
        b: 12
        bandwidth: 40000000000
        delay: 1000
      - a: 9
        b: 13
        bandwidth: 40000000000
        delay: 1000
  - fabs:
      - id: 14
        kind: Switch
      - id: 15
        kind: Switch
    racks:
      - tor:
          id: 10
          kind: Switch
        hosts:
          - id: 4
            kind: Host
          - id: 5
            kind: Host
        host2tor:
          - a: 4
            b: 10
            bandwidth: 10000000000
            delay: 1000
          - a: 5
            b: 10
            bandwidth: 10000000000
            delay: 1000
      - tor:
          id: 11
          kind: Switch
        hosts:
          - id: 6
            kind: Host
          - id: 7
            kind: Host
        host2tor:
          - a: 6
            b: 11
            bandwidth: 10000000000
            delay: 1000
          - a: 7
            b: 11
            bandwidth: 10000000000
            delay: 1000
    tor2fab:
      - a: 10
        b: 14
        bandwidth: 40000000000
        delay: 1000
      - a: 10
        b: 15
        bandwidth: 40000000000
        delay: 1000
      - a: 11
        b: 14
        bandwidth: 40000000000
        delay: 1000
      - a: 11
        b: 15
        bandwidth: 40000000000
        delay: 1000
fab2spine:
  - a: 12
    b: 16
    bandwidth: 40000000000
    delay: 1000
  - a: 12
    b: 17
    bandwidth: 40000000000
    delay: 1000
  - a: 13
    b: 18
    bandwidth: 40000000000
    delay: 1000
  - a: 13
    b: 19
    bandwidth: 40000000000
    delay: 1000
  - a: 14
    b: 16
    bandwidth: 40000000000
    delay: 1000
  - a: 14
    b: 17
    bandwidth: 40000000000
    delay: 1000
  - a: 15
    b: 18
    bandwidth: 40000000000
    delay: 1000
  - a: 15
    b: 19
    bandwidth: 40000000000
    delay: 1000

